security/pkg/k8s/controller/workloadsecret.go

```diff
@@ -441,6 +467,20 @@ func (sc *SecretController) refreshSecret(scrt *v1.Secret) error {
 	scrt.Data[PrivateKeyID] = key
 	scrt.Data[RootCertID] = sc.ca.GetCAKeyCertBundle().GetRootCertPem()
 
-	_, err = sc.core.Secrets(namespace).Update(scrt)
+	// Update secret with retry to mitigate transient failures.
+	for i := 0; i < secretOperationRetry; i++ {
+		_, err = sc.core.Secrets(namespace).Update(scrt)
+		// IsConflict may mean an optimistic lock error, i.e., the secret has been
+		// updated by other citadel instances since last read. In this case, we
```

Comment:
Interesting, I read the link below from oreily. Couple of questions.

- It's not necessarily citadel reading the secrets, but instead the kubelet mount doing the secret mounts. right? In that case, based on my understanding of the `Opportunistic concurrency` section, it's the apiserver/k8s client library abort the operation, rather the caller here. so we should retry, not return
- The reference is a book. Do we know what's the k8s version requirement, and any official k8s doc/code refernece will be more update-to-date.

Reply:
sorry, ignore the first point. The chapter there does mention about the concurrent w/ and that's what citadel multi instance case here.

maybe i don't have enough context, but then I'm confused, in our case, even previous writes are lost, does it matter? all we care about is in the end, the secrets citadel generates contains valid key/cert/root, right?

In another words, even without retry, replicated citadel instance, without this code, it should still be one succeed, and others return error, right? We use the same k8s-client methods, `sc.core.Secrets(namespace).Update(scrt)`.

Reply:
Yes, the retry is actually for nonConflict error. 

And this is generally supported by ApiServer

Reply:
This PR separates conflict errors from nonConflict ones for debugging purpose. Will have follow up PR the address remaining issues as mentioned at top.